src/main/kotlin/io/libp2p/core/multiformats/MultiaddrDns.kt

```diff
@@ -0,0 +1,116 @@
+package io.libp2p.core.multiformats
+
+import java.net.Inet4Address
+import java.net.Inet6Address
+import java.net.InetAddress
+import java.net.UnknownHostException
+
+class MultiaddrDns {
+    interface Resolver {
+        fun resolveDns4(hostname: String): List<Multiaddr>
+        fun resolveDns6(hostname: String): List<Multiaddr>
+    }
+
+    companion object {
+        private val dnsProtocols = arrayOf(Protocol.DNS4, Protocol.DNS6, Protocol.DNSADDR)
+
+        fun resolve(addr: Multiaddr, resolver: Resolver = DefaultResolver): List<Multiaddr> {
```

Comment:
I would consider asynchronous invocation here (aka `CompletableFuture`) since the DNS resolution may take a lot of time.
It also may make sense to check what non-blocking DNS resolve solutions exist to make it possible to resolve DNS in parallel without using a lot of threads

Reply:
This is something to think about, yes.

There is further work to do around DNS - the dnsaddr protocol, and also how we expose this as a configuration item (eg through the DSL) - so I'll come back and address that then.